core/destroy.go

```diff
@@ -1,15 +1,213 @@
 package core
 
 import (
+	"bufio"
 	"context"
+	"errors"
 	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
 
+	"github.com/charmbracelet/lipgloss"
 	"github.com/charmbracelet/log"
+	containerlablabels "github.com/srl-labs/containerlab/labels"
+	"github.com/srl-labs/containerlab/links"
+	"github.com/srl-labs/containerlab/nodes"
+	"github.com/srl-labs/containerlab/runtime"
 	"github.com/srl-labs/containerlab/runtime/ignite"
+	"github.com/srl-labs/containerlab/types"
+	"github.com/srl-labs/containerlab/utils"
 )
 
-// Destroy the given topology.
-func (c *CLab) Destroy(ctx context.Context, maxWorkers uint, keepMgmtNet bool) error {
+func (c *CLab) Destroy(ctx context.Context, options ...DestroyOption) (err error) {
+	opts := NewDestroyOptions()
+
+	for _, opt := range options {
+		opt(opts)
+	}
+
+	var containers []runtime.GenericContainer
+
+	if opts.all {
+		containers, err = c.ListContainers(ctx)
+	} else if c.TopoPaths.TopologyFilenameAbsPath() != "" {
+		containers, err = c.ListNodesContainersIgnoreNotFound(ctx)
+	} else {
+		var listOpts []ListOption
+
+		if c.Config.Name != "" {
+			listOpts = []ListOption{WithListLabName(c.Config.Name)}
+		} else {
+			listOpts = []ListOption{WithListContainerlabLabelExists()}
+		}
+
+		containers, err = c.ListContainers(ctx, listOpts...)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	// topo will hold the reference to the topology file
+	// as the key and the respective lab directory as the referenced value
+	topos := map[string]string{}
+
+	for _, container := range containers {
+		topoFile, ok := container.Labels[containerlablabels.TopoFile]
+		if !ok {
+			continue
+		}
+
+		topos[topoFile] = filepath.Dir(container.Labels[containerlablabels.NodeLabDir])
+	}
+
+	defer func() {
```

Comment:
rather than having cleanup take place above and below this point just deferring it all in one shot here. error should still be returned because of the named return bits